velox/functions/sparksql/specialforms/DecimalRound.cpp

```diff
@@ -0,0 +1,200 @@
+  if (inputType->isShortDecimal()) {
+    if (resultType->isShortDecimal()) {
+      return std::make_shared<DecimalRoundFunction<int64_t, int64_t>>(
+          scale, inputPrecision, inputScale, resultPrecision, resultScale);
+    } else {
+      return std::make_shared<DecimalRoundFunction<int128_t, int64_t>>(
+          scale, inputPrecision, inputScale, resultPrecision, resultScale);
+    }
+  } else {
+    if (resultType->isShortDecimal()) {
+      return std::make_shared<DecimalRoundFunction<int64_t, int128_t>>(
+          scale, inputPrecision, inputScale, resultPrecision, resultScale);
+    } else {
+      return std::make_shared<DecimalRoundFunction<int128_t, int128_t>>(
+          scale, inputPrecision, inputScale, resultPrecision, resultScale);
+    }
+  }
+}
+}; // namespace
+
+std::pair<uint8_t, uint8_t>
+getResultPrecisionScale(uint8_t precision, uint8_t scale, int32_t roundScale) {
+  // After rounding we may need one more digit in the integral part.
+  int32_t integralLeastNumDigits = precision - scale + 1;
+  if (roundScale < 0) {
+    // Negative scale means we need to adjust `-scale` number of digits before
+    // the decimal point, which means we need at least `-scale + 1` digits after
+    // rounding, and the result scale is 0.
+    auto newPrecision = std::max(integralLeastNumDigits, -roundScale + 1);
+    // We have to accept the risk of overflow as we can't exceed the max
+    // precision.
+    return {std::min(newPrecision, 38), 0};
+  } else {
+    uint8_t newScale = std::min((int32_t)scale, roundScale);
+    // We have to accept the risk of overflow as we can't exceed the max
+    // precision.
+    return {std::min(integralLeastNumDigits + newScale, 38), newScale};
+  }
+}
+
+TypePtr DecimalRoundCallToSpecialForm::resolveType(
+    const std::vector<TypePtr>& argTypes) {
+  VELOX_FAIL("Decimal round function does not support type resolution.");
+}
+
+exec::ExprPtr DecimalRoundCallToSpecialForm::constructSpecialForm(
+    const TypePtr& type,
+    std::vector<exec::ExprPtr>&& compiledChildren,
+    bool trackCpuUsage,
+    const core::QueryConfig& config) {
+  VELOX_USER_CHECK_GE(
+      compiledChildren.size(),
+      1,
+      "Decimal_round expects one or two arguments.");
+  VELOX_USER_CHECK_LE(
+      compiledChildren.size(),
+      2,
+      "Decimal_round expects one or two arguments.");
+  VELOX_USER_CHECK(
+      compiledChildren[0]->type()->isDecimal(),
+      "The first argument of decimal_round should be of decimal type.");
+  int32_t scale = 0;
+  if (compiledChildren.size() > 1) {
+    VELOX_USER_CHECK_EQ(
+        compiledChildren[1]->type()->kind(),
+        TypeKind::INTEGER,
+        "The second argument of decimal_round should be of integer type.");
+    auto constantExpr =
+        std::dynamic_pointer_cast<exec::ConstantExpr>(compiledChildren[1]);
+    VELOX_USER_CHECK_NOT_NULL(constantExpr);
+    VELOX_USER_CHECK(constantExpr->value()->isConstantEncoding());
```

Comment:
check for null constant; also check for abs('scale') to be within reasonable bounds

Reply:
All intergers are acceptable as the scale.